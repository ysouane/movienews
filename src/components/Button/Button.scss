@import 'styles/variables';

.button {
  display: inline-block;
  padding: 0.3em 1.2em;
  margin: 0 0.3em 0.3em 0;
  border-radius: 2em;
  box-sizing: border-box;
  text-decoration: none;
  font-family: $font-family-base;
  font-weight: 300;
  color: #FFFFFF;
  background-color: #3782ad;
  text-align: center;
  transition: all 0.2s;

  &:hover {
    background-color: #4095c6;
  }

  @media all and (max-width:30em) {
    & {
      display: block;
      margin: 0.2em auto;
    }
  }
}