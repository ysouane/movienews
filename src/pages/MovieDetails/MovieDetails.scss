@import 'styles/variables';

.movieDetails {
  display: flex;
  flex-direction: column;
  color: $white;
  margin: 2rem;
  border: 2px solid $white;
  padding: 2rem;

  @media screen and (min-width: $breakpoints-md) {
    flex-direction: row;
  }

  &__backButtonContainer {
    margin-left: 2rem;

    @media screen and (max-width: $breakpoints-md) {
      margin-left: 0;
      display: flex;
      justify-content: center;
    }
  }

  &__image,
  &__title {
    display: block;
    object-fit: cover;
    width: 100%;
    height: 400px;
    border: 2px solid $white;

    @media screen and (min-width: $breakpoints-md) {
      width: 175px;
      height: 250px;
    }
  }

  &__title {
    background-color: $white;
    color: $black;
    display: flex;
    flex-direction: column;
    justify-content: center;
    text-align: center;
  }

  &__overview {
    margin: 1rem;

    @media screen and (min-width: $breakpoints-md) {
      width: 60%;
    }
  }
}