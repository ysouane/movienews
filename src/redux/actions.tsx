export const changeTypeOfDisplayAndUrl = (value: string) => ({
  type: "changeTypeOfDisplayAndUrl",
  payload: { value }
})
