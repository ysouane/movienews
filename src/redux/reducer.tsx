import { API_KEY, API_URL, API_VERSION } from "helpers"
import currentDate from "utils/currentDateUTCFormatted"

export default function Reducer(state: any, action: any) {
  if (action.type === "changeTypeOfDisplayAndUrl") {
    const { value } = action.payload

    if (value === "recentMovies") {
      return {
        ...state,
        typeOfDisplay: "recentMovies",
        url: `${API_URL}/${API_VERSION}/discover/movie?api_key=${API_KEY}&language=fr-Fr&sort_by=release_date.desc&release_date.lte=${currentDate}&page=1`
      }
    }

    if (value === "popularMovies") {
      return {
        ...state,
        typeOfDisplay: "popularMovies",
        url: `${API_URL}/${API_VERSION}/movie/popular?api_key=${API_KEY}&language=fr-FR&page=1`
      }
    }

    if (value === "favorites") {
      return {
        ...state,
        typeOfDisplay: "favorites"
      }
    }
  }

  return state
}
