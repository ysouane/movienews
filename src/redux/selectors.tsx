export const checkTypeOfDisplay = (state: any) => state.typeOfDisplay
export const checkUrl = (state: any) => state.url
