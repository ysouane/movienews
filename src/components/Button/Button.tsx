import React from "react"
import "./Button.scss"

const Button = ({ navigate, text, addOrRemoveIdFromLocalStorage }: any) => {
  return (
    <button
      className="button"
      onClick={
        text === "Retour" ? () => navigate(-1) : addOrRemoveIdFromLocalStorage
      }
      type="button"
    >
      {text}
    </button>
  )
}

export default Button
