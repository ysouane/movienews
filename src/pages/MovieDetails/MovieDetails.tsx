import React, { useEffect, useState } from "react"
import { useLocation, useNavigate } from "react-router-dom"
import { API_KEY, API_URL, API_VERSION } from "helpers"
import Button from "components/Button"
import "./MovieDetails.scss"

const FAVORITES_CHOICE = {
  ADD: "Ajouter aux favoris",
  REMOVE: "Supprimer des favoris"
}

const MovieDetails = () => {
  const [data, setData] = useState<any>([])
  const [favoritesText, setFavoritesText] = useState(FAVORITES_CHOICE.ADD)
  const location: any = useLocation()
  const navigate: any = useNavigate()
  const { id }: any = location.state || 0 // 0 is the default value in the case where id is null
  const urlMovieDetail = `${API_URL}/${API_VERSION}/movie/${id}?api_key=${API_KEY}&language=fr-Fr&append_to_response=videos,images`

  useEffect(() => {
    async function fetchData() {
      await fetch(urlMovieDetail)
        .then((res) => res.json())
        .then((resJson) => setData(resJson))
        .catch((error) => {
          throw error
        })
    }

    if (id) {
      fetchData()
    }
  }, [id, urlMovieDetail])

  useEffect(() => {
    const favorites = window.localStorage.movies?.split(",")

    const idInFavorites = favorites?.some(
      (favId: string) => favId === id.toString()
    )

    if (idInFavorites) {
      setFavoritesText(FAVORITES_CHOICE.REMOVE)
    }
  }, [id])

  const addOrRemoveIdFromLocalStorage = () => {
    let favorites = window.localStorage.movies

    if (favorites !== undefined) {
      favorites = window.localStorage.movies.split(",")
    } else {
      favorites = []
    }

    const idInFavorites = favorites.some(
      (favId: string) => favId === id.toString()
    )

    if (idInFavorites) {
      if (favorites.length === 1) {
        window.localStorage.removeItem("movies")
      } else {
        const newData = favorites.filter(
          (favId: string) => favId !== id.toString()
        )
        window.localStorage.movies = newData
      }

      setFavoritesText(FAVORITES_CHOICE.ADD)
    } else {
      favorites.push(id)
      window.localStorage.movies = favorites
      setFavoritesText(FAVORITES_CHOICE.REMOVE)
    }
  }

  if (data) {
    return (
      <>
        <div className="movieDetails__backButtonContainer">
          <Button navigate={navigate} text="Retour" />
        </div>
        <div className="movieDetails">
          {data.poster_path ? (
            <img
              className="movieDetails__image"
              src={`https://image.tmdb.org/t/p/w500${data.poster_path}`}
              alt={data.original_title}
            />
          ) : (
            <span className="movieDetails__title">{data.original_title}</span>
          )}

          <div className="movieDetails__overview">
            <p>
              Titre du film:{" "}
              {data.original_title}
            </p>
            <p>
              Date de sortie:{" "}
              {data.release_date}
            </p>
            <p>
              Résumé:{" "}
              {data.overview ? data.overview : "Pas de résumé disponible"}
            </p>

            <Button
              navigate={navigate}
              text={favoritesText}
              addOrRemoveIdFromLocalStorage={addOrRemoveIdFromLocalStorage}
            />
          </div>
        </div>
      </>
    )
  }

  return (
    <div>
      <Button navigate={navigate} />
      <p>Désolé pas de détails pour ce film</p>
    </div>
  )
}

export default MovieDetails
