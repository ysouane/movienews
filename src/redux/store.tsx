import { createStore } from "redux"
import { API_KEY, API_URL, API_VERSION } from "helpers"
import currentDate from "utils/currentDateUTCFormatted"
import Reducer from "./reducer"

export const initialState: any = {
  typeOfDisplay: "recentMovies",
  url: `${API_URL}/${API_VERSION}/discover/movie?api_key=${API_KEY}&language=fr-Fr&sort_by=release_date.desc&release_date.lte=${currentDate}&page=1`,
  favorites: []
}

export const store = createStore(Reducer, initialState)
