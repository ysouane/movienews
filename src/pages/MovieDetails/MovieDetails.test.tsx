import React from "react"
import { render } from "@testing-library/react"
import renderer from "react-test-renderer"
import MovieDetails from "./MovieDetails"
import Button from "components/Button"
import { Provider } from "react-redux"
import { store } from "redux/store"
import { BrowserRouter } from "react-router-dom"

describe("MovieDetails page", () => {
  it("Should return without error", () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <MovieDetails />
        </Provider>
      </BrowserRouter>
    )
  })
})

it("renders button component correctly", () => {
  const tree = renderer.create(<Button />).toJSON()
  expect(tree).toMatchSnapshot()
})
