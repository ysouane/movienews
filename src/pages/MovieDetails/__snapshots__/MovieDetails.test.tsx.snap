// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button component correctly 1`] = `
<button
  className="button"
  type="button"
/>
`;
